### ISSUE TITLE
x64: bug in getModRM()

### ISSUE THREAD

**Issue by LRGH** [bdcht/amoco#48]
``` python
>>> from amoco.arch.x64 import cpu_x64
>>> i=cpu_x64.disassemble('\x48\x8b\x04\xc5\0\0\0\0')
>>> str(i.operands[1])
'M64cs(((rax*0x8)+rip))'
```

but it should be `M64cs((rax*0x8))`

There is also

``` python
>>> from amoco.arch.x64 import cpu_x64
>>> i=cpu_x64.disassemble('\x64\x48\x8b\x04\x25\x28\0\0\0')
>>> str(i.operands[1])
'M64fs(rip+40)'
```

which should be `M64fs(40)`.

The proposed patch below addresses these problems.

``` python
--- a/amoco/arch/x64/utils.py
+++ b/amoco/arch/x64/utils.py
@@ -96,12 +96,16 @@ def getModRM(obj,Mod,RM,data):

     # check [disp16/32] case:
     if (b is env.rbp or b is env.r13) and Mod==0:
-        b=env.rip
-        if seg is '': seg = env.cs
-        Mod = 0b10
-    if (b is env.bp) and Mod==0:
-        b=env.cst(0,adrsz)
+        if s is 0 and seg is '':
+            seg = env.cs
+            bs = env.rip
+        else:
+            bs = s + env.cst(0,adrsz)
         Mod = 0b10
+    elif s is 0:
+        bs = b
+    else:
+        bs = b + s
     # now read displacement bytes:
     if Mod==0b00:
         d = 0
@@ -119,7 +123,6 @@ def getModRM(obj,Mod,RM,data):
         obj.bytes += pack(d)
         data = data[immsz:data.size]
         d = d.int(-1)
-    bs = b+s
     if bs._is_cst and bs.v==0x0:
         bs.v = d
         bs.size = adrsz
```

Note that I wrote this patch such that `bs = b + s` appears only when both are non-zero.
It allows me to replace this addition by the following piece of code, that I need because I want amoco to remember whether the binary conained `8b 04 18` (aka. `mov (%rax,%rbx), %eax`) or `8b 04 03` (aka. `mov (%rbx,%rax), %eax`).

``` python
        # Instead of doing bs = b+s, which will reorder arguments, we do
        # the addition manually, and change 'prop' so the many future calls
        # to 'simplify' does not reorder the arguments
        from amoco.cas import expressions
        bs = expressions.op('+', b, s)
        bs.prop |= 16
```

**Comment by LRGH** [comment]
Note that there are probably other issues.
In `getModRM()` there are many tests `adrsz!=16` but `adrsz` is never set to `16`!

**Comment by bdcht** [comment]
Indeed, x64 arch still lacks testings ! 
I'm currently fixing the x64 getModRM which was basically copied from x86...obviously many x64-specific cases were not correctly decoded (mostly related to the rip-relative addressing mode introduced for x64). I'll add a dedicated test_arch_x64.py as well...

**Comment by LRGH** [comment]
Yes, my proposed patch above is not sufficient. `8b 2c 25 00 00 00 00` still decodes as `movl (%rip), %ebp` instead of `movl 0, %ebp`.

Now I propose `if s is 0 and seg is '' and RM!=0b100:` instead of `if s is 0 and seg is ''`.

**Comment by bdcht** [comment]
fixed by a6535e5 (v2.4.5), including for last comment.

**Comment by LRGH** [comment]
Thank you.
https://github.com/bdcht/amoco/issues/53 is a typo in this patch.
